---
title: "scRNABatchQC Report"
output: 
  html_document:
    toc: true
    toc_depth: 2
    number_sections: true
params:
  data: plotData
  displayFigure: TRUE
---

<style type="text/css">
.figure p.caption {
  font-size: 1.5em;
}
.main-container {
  max-width: 90%;
  margin-left: auto;
  margin-right: auto;
}
</style>

```{r setup, include=FALSE}
library(RColorBrewer)
library(knitr)
library(kableExtra)
library(reshape2)
library(ggplot2)
library(data.table)
library(DT)
library(RCurl)
library(htmltools)

knitr::opts_chunk$set(echo = TRUE, autodep=TRUE, cache.comments=FALSE, message=FALSE, warning=FALSE, dpi=300)

addLinkTag<-function(text,link) {
  result<-paste0("<a href='",link,"' target='_blank'>",text,"</a>")
  return(result)
}

figRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("figcap.prefix"), 
           sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight"), trunk.eval=TRUE) {
    if(trunk.eval){ 
      i <- which(names(tag) == label)
      if (length(i) == 0) {
        i <- length(tag) + 1
        tag <<- c(tag, i)
        names(tag)[length(tag)] <<- label
        used <<- c(used, FALSE)
        names(used)[length(used)] <<- label
        created <<- c(created, FALSE)
        names(created)[length(created)] <<- label
      }
      if (!missing(caption)) {
        created[label] <<- TRUE
        paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
               " ", caption)
      } else {
        used[label] <<- TRUE
        paste(prefix, tag[label])
      }
    }
  }
})
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")

tabRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("tabcap.prefix"), 
           sep = options("tabcap.sep"), prefix.highlight = options("tabcap.prefix.highlight"), trunk.eval=TRUE) {
    if(trunk.eval){
      i <- which(names(tag) == label)
      if (length(i) == 0) {
        i <- length(tag) + 1
        tag <<- c(tag, i)
        names(tag)[length(tag)] <<- label
        used <<- c(used, FALSE)
        names(used)[length(used)] <<- label
        created <<- c(created, FALSE)
        names(created)[length(created)] <<- label
      }
      if (!missing(caption)) {
        created[label] <<- TRUE
        paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
              " ", caption)
      } else {
        used[label] <<- TRUE
        paste(prefix, tag[label])
      }
    }
  }
})
options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")

getHeatmapHeight<-function(hdata){
  max(10, nrow(hdata) / 10)
}

isHeatmapFigure<-function(datam){
  if (is.null(datam)){
    return(FALSE)
  }
  return (ncol(datam)>1 & nrow(datam)>1)
}

getHeatmapCaption<-function(datam, feature){
  if (isHeatmapFigure(datam)){
    return (figRef(feature))
  }else{
    return (tabRef(feature))
  }
}

plotHeatmap<-function(datam,feature=c("hvg","hvgPath","PCg","PCgPath","diffg","diffgPath")){
  keytitle<-switch(feature, "hvg"="zval","hvgPath"="-logFDR","PCg"="logFC","PCgPath"="-logFDR","diffg"="logFC","diffgPath"="-logFDR")
  maintitle<-switch(feature, "hvg"="Highly variable genes","hvgPath"="Pathways enriched in HVGs","PCg"="PC-related genes","PCPath"="Pathways enriched in PC-related genes","diffg"="Differentially expressed genes","diffgPath"="Pathways enriched in DEGs")
  if (isHeatmapFigure(datam)){
    if (feature== "hvg" | feature== "PCg" | feature=="diffg"){ 
      colv=heatmapColors
	  } 
    else {
      colv=heatmapPathwayColors
    }
    heatmap.2(datam, margins = heatmapMargins,  keysize=heatmapKeysize, col=colv,main=maintitle,key.title=keytitle,key.xlab="",key.ylab="",cexCol=1)
  } 
  else {
    warning("there is only one dataset/comparison or one gene/pathway detected and heatmap figure is not generated")
	  print(kable(as.matrix(datam),caption=tabRef(feature, maintitle),col.names=paste0(colnames(datam),"(",keytitle,")")) %>% 
	          kable_styling() %>%
	          htmltools::HTML())
	}
}

```

```{r data, include=FALSE, cache=FALSE}
#library(scRNABatchQC)
#load("test.rdata")
plotData<-params$data
allsces <- plotData$sces

sces<-list()
unvalid<-list()

for (p in names(allsces)) {
  sce<-allsces[[p]]
  if(sce@metadata$valid){
    sces[p] = sce
  }else{
    unvalid[p] = sce
  }
}

if(length(allsces) < 9){
  scolors=brewer.pal(length(allsces), "Set1")
}else{
  scolors=rainbow(length(allsces))
}
names(scolors)<-names(allsces)

hasValid<-length(sces) > 0
hasNotValid<-length(unvalid) > 0

if(hasValid){
  hvgBiologicalSimilarity <- .getBiologicalSimilarity(sces, objectName = "hvg", valueName = "zval")
  hvgPathways <- .getMultiplePathways(sces, metaObjectName = "hvgPathway")
  pc1geneBiologicalSimilarity <- .getBiologicalSimilarity(sces, objectName = "pc1genes", valueName = "logFC")
  pc1Pathways <- .getMultiplePathways(sces, metaObjectName = "pc1Pathway")
  
  hasComparison<-length(sces) > 1

  heatmapKeysize<-1
  heatmapColors<-bluered(75)
  heatmapPathwayColors<-colorpanel(75,low="white",high="red")
  heatmapMargins<-c(10,15)
  
  hasPathway<-!is.null(hvgPathways)
  
  hasPC1Pathway<-!is.null(pc1Pathways)
  
  hasDiffGenes<-!is.null(plotData$scesMerge@metadata$diffFC$genes)
  
  hasDiffPathway<-!is.null(plotData$scesMerge@metadata$diffFC$pathways)
}else{
  hasComparison<-FALSE
  hasPathway<-FALSE
  hasPC1Pathway<-FALSE
  hasDiffGenes<-FALSE
  hasDiffPathway<-FALSE
}

```

# Overview

**scRNABatchQC provides both metrics and diagnostic graphics to assess the similarity/difference on technical factors, expression profiles, and biological features across multiple scRNA-seq experiments, which helps identify potential batch effects. scRNABatchQC report includes five sections:**

* _**QC Summary**_ - Overview of quality control in each scRNA-seq experiment. 
* _**Technical View**_ - Technical features across multiple scRNA-seq experiments. 
* _**Biological View**_ - Biological features across multiple scRNA-seq experiments.
* _**Expression Similarity**_ - Expression similarity across multiple scRNA-seq experiments. 
* _**Pairwise Difference**_ - Genes and pathways that drive the pairwise difference across multiple scRNA-seq experiments.

# QC Summary

```{r tableSummary, echo=FALSE, results="asis"}
curTable<-plotData$tableSummary
print(kable(curTable, caption=tabRef("tableSummary", "Summary of Quality Metrics."), row.names=FALSE) %>% 
      kable_styling() %>% 
		  row_spec(which(curTable$Valid=="FALSE"), color = "black", background = "bisque") %>% 
		  htmltools::HTML())

```

`r tabRef("tableSummary")` contains a summary of quality metrics generated by the scRNABatchQC package. 

A short description of Table 1 metrics is provided below: 

* _**EID**_ - The name of the experiment
* _**Count**_ - The total number of counts in the experiment 
* _**Cell**_ - The total number of cells in the experiment
* _**Gene**_ - The total number of genes in the experiment
* _**R-Count**_ - The range of the total number of counts in a cell across all cells [min-median-max]. 
* _**R-Gene**_ - The range of the total number of genes identified in a cell across all cells [min-median-max] 
* _**mtRNA**_ - The maximum percentage of reads mapped to mitochondria-encoded genes in all cells 
* _**rRNA**_ - The maximum percentage of reads mapped to genes encoded for ribosomal proteins in all cells
* _**F-Count**_ - The number of cells filtered by the total number of counts 
* _**C-Count**_ - The cutoff to remove the cell if the total counts of the cell is less than C-Count
* _**F-Gene**_ - The number of cells filtered by the total number of detected genes
* _**C-Gene**_ - The cutoff to remove the cell if the total genes of the cell is less than C-Gene
* _**F-mt**_ - The number of cells filtered by the mtRNA percentage 
* _**C-mt**_ - The cutoff to remove the cell if the percentage of mtRNA is larger than C-mt
* _**F**_ - The total number of cells filtered by either counts, genes, or mtRNA criteria
* _**V-Count**_ - The number of cell passed filtering
* _**Valid**_ - If the sample was unvalid due to extremly low cell count (less than 10)
