# Biological View

**This section presents the similarity/difference on the highly variable genes and genes differentially expressed along a specific principal component (PCind, default: the first PC) between scRNA-seq experiments. scRNABatchQC calculates the pairwise similarity of highly variable genes and PC-related genes and their enriched pathways across experiments. Two experiments with similar highly variable genes or PC-related gens are either derived from the similar cell populations, or subject to same technical bias. Pathway enrichment analysis will help discriminate real biological effect from just technical noise.**

<hr>

```{r plotBiologicalSimilarityHVGenes, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("hvg", "Heatmap of top50 highly variable genes", trunk.eval= isHeatmapFigure(hvgBiologicalSimilarity)), fig.height=getHeatmapHeight(hvgBiologicalSimilarity), fig.width=10}
plotHeatmap(hvgBiologicalSimilarity,"hvg")
```

`r getHeatmapCaption(hvgBiologicalSimilarity,"hvg")` shows the z-score of top 50 HVGs in each experiment and similarity across experiments if there are multiple samples. scRNABatchQC first selects the top 50 genes with highest and significant biological variance in each experiment (FDR<0.01) and then clusters experiments based on the similarity on highly variable genes.

<hr>


```{r plotPathwaySimilarityHVG, eval=hasPathway, echo=FALSE, warning=FALSE, message=FALSE, error=FALSE, results='asis', fig.align="center", fig.cap=figRef("hvgPath", "Heatmap of enriched pathways for highly variable genes", trunk.eval=hasPathway & isHeatmapFigure(hvgPathways)), fig.height=getHeatmapHeight(hvgPathways), cache=FALSE, fig.width=10}
plotHeatmap(hvgPathways,"hvgPath")
cat(getHeatmapCaption(hvgPathways,"hvgPath")," shows the -log10 FDR for enriched pathways for HVGs in each experiment and pathway similarity across experiments if there are multiple samples.")

```
`r getHeatmapCaption(hvgPathways,"hvgPath")` shows the -log10 FDR for enriched pathways for HVGs in each experiment and pathway similarity across experiments if there are multiple samples.


<hr>

```{r plotBiologicalSimilarityPC1Genes, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("PCg", "Heatmap of top50 PC-related genes", trunk.eval=isHeatmapFigure(pc1geneBiologicalSimilarity)), fig.height=10, fig.width=10}
plotHeatmap(pc1geneBiologicalSimilarity,"PCg")
```

`r getHeatmapCaption(pc1geneBiologicalSimilarity,"PCg")` shows differential expression of top 50 PC-related genes in each experiment and similarity across experiments if there are multiple samples. scRNABatchQC selects the top 50 differentially expressed genes along the specific principle component (PCind, default: the first PC) in each experiment (FDR<0.01).

<hr>

```{r plotPathwaySimilarityPC1Gene, eval=hasPC1Pathway, echo=FALSE, warning=FALSE, message=FALSE, error=FALSE, results='asis', fig.align="center", fig.cap=figRef("PCgPath", "Heatmap of enriched pathways for PC-related genes", trunk.eval=hasPC1Pathway & isHeatmapFigure(pc1Pathways)), fig.height=8, cache=FALSE, fig.width=10}
plotHeatmap(pc1Pathways,"PCgPath")
cat(getHeatmapCaption(pc1Pathways,"PCgPath"), "shows the -log10 FDR for enriched pathways for PC-related genes in each experiment and pathway similarity across experiments if there are multiple samples.")

```

<hr>
