---
title: "QC for single cell RNA-seq data"
output: 
  html_document:
    toc: true
    toc_float: true
params:
  data: sces
  all: scesall
  organ: organism
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(RColorBrewer)
library(knitr)
library(reshape2)
library(ggplot2)
library(data.table)
library(DT)
library(RCurl)
library(htmltools)

knitr::opts_chunk$set(echo = TRUE)

addLinkTag<-function(text,link) {
  result<-paste0("<a href='",link,"' target='_blank'>",text,"</a>")
  return(result)
}

figRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("figcap.prefix"), 
           sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight")) {
    i <- which(names(tag) == label)
    if (length(i) == 0) {
      i <- length(tag) + 1
      tag <<- c(tag, i)
      names(tag)[length(tag)] <<- label
      used <<- c(used, FALSE)
      names(used)[length(used)] <<- label
      created <<- c(created, FALSE)
      names(created)[length(created)] <<- label
    }
    if (!missing(caption)) {
      created[label] <<- TRUE
      paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
             " ", caption)
    } else {
      used[label] <<- TRUE
      paste(prefix, tag[label])
    }
  }
})
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")

tabRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("tabcap.prefix"), 
           sep = options("tabcap.sep"), prefix.highlight = options("tabcap.prefix.highlight")) {
    i <- which(names(tag) == label)
    if (length(i) == 0) {
      i <- length(tag) + 1
      tag <<- c(tag, i)
      names(tag)[length(tag)] <<- label
      used <<- c(used, FALSE)
      names(used)[length(used)] <<- label
      created <<- c(created, FALSE)
      names(created)[length(created)] <<- label
    }
    if (!missing(caption)) {
      created[label] <<- TRUE
      paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
             " ", caption)
    } else {
      used[label] <<- TRUE
      paste(prefix, tag[label])
    }
  }
})
options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")
```

```{r data, include=FALSE, cache=FALSE}
sces <- params$data
scesall <- params$all
organism <- params$organ

hvgPathways<-.getMultiplePathway(sces, metaObjectName="hvgPathway")
pc1Pathways<-.getMultiplePathway(sces, metaObjectName="pc1Pathway")

diffFC <- .getDiffGenes(scesall, organism = organism,  FDR = 0.01, geneNo = 50)

if(length(sces) < 9){
  scolors=brewer.pal(length(sces), "Set1")
}else{
  scolors=rainbow(length(sces))
}
names(scolors)<-names(sces)
```

### 1. Technical features

```{r dis_total_counts, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("density_total_counts", "Total count distribution")}
print(plotDensity(sces,"total_counts", "Total count", scolors))
```

```{r dis_total_feature, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("density_total_features", "Total feature distribution")}
print(plotDensity(sces,"total_features", "Total feature", scolors))
```

```{r pct_counts_Mt, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("density_pct_counts_Mt", "Mitochondrial count distribution")}
print(plotDensity(sces,"pct_counts_Mt", "pct_counts_Mt", scolors))
```

```{r GeneCountDistribution, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("genesCountdistribution", "Top 500 Genes Count Distribution")}
print(plotGeneCountDistribution(sces, scolors))
```

```{r aveCountVSdetectRate, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("aveCountVSdetectRate", "Average expression and detection rate of genes")}
print(plotAveCountVSdetectRate(sces, scolors))
```

```{r varianceTrend, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("VarianceTrend", "Mean-variance trend")}
print(plotVarianceTrend(sces, scolors))
```

```{r log10_total_counts, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("log10_total_counts", "log10(total counts)")}
print(plotMultiSamplesOneExplanatoryVariables(sces, scolors, "log10_total_counts"))
```

```{r log10_total_features, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("log10_total_features", "log10(total features)")}
print(plotMultiSamplesOneExplanatoryVariables(sces, scolors, "log10_total_features"))
```

```{r log10_total_counts_Mt, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("log10_total_counts_Mt", "log10(total mitochondrial count)")}
print(plotMultiSamplesOneExplanatoryVariables(sces, scolors, "log10_total_counts_Mt"))
```

### 2. Sample expression similarity 

```{r samplesimilarity, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", warning=FALSE, fig.cap=figRef("samplesimilarity", "Sample Similarity")}
plotSampleSimilarity(sces)
```

```{r plotPCA, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotPCA", "PCA")}
print(plotAllPCA(scesall, scolors))
```

```{r plotTSNE, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotTSNE", "TSNE")}
print(plotAllTSNE(scesall, scolors))
```

### 3. Biological Features similarity

```{r biologicalSimilarityHVGenes, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("biologicalSimilarityHVGenes", "Top 50 highly variable genes with FDR < 0.01"), fig.height=10}
plotBiologicalSimilarity(sces, objectName="hvg", filterName="FDR", valueName="bio", cexRow = 0.5, cexCol = 0.5)
```

```{r pathwaySimilarityHVG, echo=FALSE, warning=FALSE, message=FALSE, error=FALSE, results='asis', fig.align="center", fig.cap=figRef("pathwaySimilarityHVG", "Pathway similarity based on highly variable genes with FDR < 0.01"), fig.height=10, cache=FALSE}
heatmap.2(hvgPathways, margins = c(5, 10), cexRow = 0.5)
```

```{r biologicalSimilarityPC1Genes, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("biologicalSimilarityPC1Genes", "Top 50 PC1 genes with FDR < 0.01"), fig.height=10}
plotBiologicalSimilarity(sces, objectName="pc1genes", filterName="adj.P.Val", valueName="logFC", cexRow = 0.5, cexCol = 0.5)
```

```{r pathwaySimilarityPC1Genes, echo=FALSE, warning=FALSE, message=FALSE, error=FALSE, results='asis', fig.align="center", fig.cap=figRef("pathwaySimilarityPC1Gene", "Pathway similarity based on PC1 genes with FDR < 0.01"), fig.height=10, cache=FALSE}
heatmap.2(pc1Pathways, margins = c(5, 10), cexRow = 0.5)
```

### 4. Pairwise difference

```{r plotPC, echo=FALSE, warning=FALSE, message=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotPairComp", "Differentially Expressed Genes"), fig.height=8}
heatmap.2(as.matrix(diffFC$genes), cexRow = 0.6, cexCol = 0.6, margins = c(5, 10))
```
