---
title: "QC for single cell RNA-seq data"
output: 
  html_document:
    toc: true
    toc_float: true
params:
  data: sces
  all: sceall
---

```{r setup, include=FALSE}

library(RColorBrewer)
library(knitr)
library(reshape2)
library(ggplot2)
library(gplots)
library(data.table)
library(DT)
library(RCurl)
library(htmltools)
library(gplots)
library(WebGestaltR)

knitr::opts_chunk$set(echo = TRUE, cache.lazy = FALSE)

addLinkTag<-function(text,link) {
  result<-paste0("<a href='",link,"' target='_blank'>",text,"</a>")
  return(result)
}

figRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("figcap.prefix"), 
           sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight")) {
    i <- which(names(tag) == label)
    if (length(i) == 0) {
      i <- length(tag) + 1
      tag <<- c(tag, i)
      names(tag)[length(tag)] <<- label
      used <<- c(used, FALSE)
      names(used)[length(used)] <<- label
      created <<- c(created, FALSE)
      names(created)[length(created)] <<- label
    }
    if (!missing(caption)) {
      created[label] <<- TRUE
      paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
             " ", caption)
    } else {
      used[label] <<- TRUE
      paste(prefix, tag[label])
    }
  }
})
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")

tabRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("tabcap.prefix"), 
           sep = options("tabcap.sep"), prefix.highlight = options("tabcap.prefix.highlight")) {
    i <- which(names(tag) == label)
    if (length(i) == 0) {
      i <- length(tag) + 1
      tag <<- c(tag, i)
      names(tag)[length(tag)] <<- label
      used <<- c(used, FALSE)
      names(used)[length(used)] <<- label
      created <<- c(created, FALSE)
      names(created)[length(created)] <<- label
    }
    if (!missing(caption)) {
      created[label] <<- TRUE
      paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
             " ", caption)
    } else {
      used[label] <<- TRUE
      paste(prefix, tag[label])
    }
  }
})
options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")

source("prepareSCRNAData.R")
source("prepareSCRNADataSet.R")
source("plotFunctions.R")
source("BiologicalSimilarityFunctions.R")

```

```{r data, include=FALSE, cache=TRUE}
  
#sces<-prepareSCRNADataSet(sampleTable)

sces <- params$data
sceall <- params$all

if(length(sces) < 9){
  scolors <- brewer.pal(length(sces), "Set1")
}else{
  scolors <- rainbow(length(sces))
}
names(scolors) <- names(sces)


```

## QC for single cell RNA-seq data

### 1. Technical features

```{r dis_total_counts,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("density_total_counts", "Distribution of total counts")}
print(plotDensity(sces,"total_counts", "Total counts", scolors))
```

```

```{r pct_counts_Mt,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("density_pct_counts_Mt", "Distribution of mitochondrial RNA")}
print(plotDensity(sces,"pct_counts_Mt", "PCT mitochondrial counts", scolors))
```

```

```{r AveCountVSNumberOfCells,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("AveCountVSNumberOfCells", "Average expression (x) and number of cells detect this gene (Y)")}
print(plotAveCountVSNumberOfCells(sces, scolors))
```

```


```



```

```{r keggHighVariableGenes,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("biologicalSimilarityHighVariableGenes", "Biological variation of most varied genes"), fig.height=10}

wg<-WebGestaltR(enrichMethod="ORA",organism="mmusculus",
            enrichDatabase="pathway_KEGG",interestGene=hvgenes,
            interestGeneType="genesymbol",referenceSet="genome",
            is.output=FALSE)


```{r biologicalSimilarityPC1Genes,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("biologicalSimilarityPC1Genes", "Biological variation of PC1 genes"), fig.height=10}
pc1genes<-plotBiologicalSimilarity(sces, objectName="pc1genes", filterName="adj.P.Val", valueName="logFC")
wg<-WebGestaltR(enrichMethod="ORA",organism="mmusculus",
            enrichDatabase="pathway_KEGG",interestGene=pc1genes,
            interestGeneType="genesymbol",referenceSet="genome",
            is.output=FALSE)
```

### 2. Sample expression similarity

#### Mean expression similarity

```{r plotsamplesimilarity, echo=FALSE, results='asis', fig.align="center", warning=FALSE, fig.cap=figRef("SampleSimilarity", "Sample Similarity")}
print(plotSampleSimilarity(sces))
```

#### PCA plot

```{r plotPCA, echo=FALSE, results='asis', fig.align="center", fig.cap=figRef("PCA", "PCA")}
print(plotAllPCA(sceall, scolors))
```

#### TSNE plot

```{r plotTSNE, echo=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotTSNE", "TSNE")}
print(plotAllTSNE(sceall, scolors))
```

### 3. Biological Features similarity

#### Highly variable genes (select the top 50 genes (adjustable) with FDR<0.01)

```{r plotHVG, echo=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotHVG", "Highly variable genes")}
hvgbio <- getHighVarGenes(sces, FDR = 0.01, geneNo = 50)
heatmap.2(hvgbio, margins = c(2, 10), cexRow = 0.5)
```

#### PC1 genes (select the top 50 genes (adjustable) with FDR<0.01)

```{r plotPC1gene, echo=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotPC1gene", "PC1 genes")}
pcgFC <- getTopPC1genes(sces, FDR = 0.01, geneNo = 50)
heatmap.2(abs(pcgFC), margins = c(2, 10), cexRow = 0.5)
```

#### Pairwise comparsion

```{r plotPC, echo=FALSE, results='asis', fig.align="center", fig.cap=figRef("plotPairComp", "Differentially Expressed Genes")}
diffFC <- getDiffGenes(sceall, FDR = 0.01, geneNo = 50)
heatmap.2(diffFC, cexRow = 0.6, cexCol = 0.6, margins = c(5, 10))
```
