---
title: "QC for single cell RNA-seq data"
output: html_document
params:
  data: dat
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)

library(RColorBrewer)
library(knitr)
library(reshape2)
library(ggplot2)
library(data.table)
library(DT)
library(RCurl)
library(htmltools)
library(gplots)
library(WebGestaltR)

knitr::opts_chunk$set(echo = TRUE)

addLinkTag<-function(text,link) {
  result<-paste0("<a href='",link,"' target='_blank'>",text,"</a>")
  return(result)
}

figRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("figcap.prefix"), 
           sep = options("figcap.sep"), prefix.highlight = options("figcap.prefix.highlight")) {
    i <- which(names(tag) == label)
    if (length(i) == 0) {
      i <- length(tag) + 1
      tag <<- c(tag, i)
      names(tag)[length(tag)] <<- label
      used <<- c(used, FALSE)
      names(used)[length(used)] <<- label
      created <<- c(created, FALSE)
      names(created)[length(created)] <<- label
    }
    if (!missing(caption)) {
      created[label] <<- TRUE
      paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
             " ", caption)
    } else {
      used[label] <<- TRUE
      paste(prefix, tag[label])
    }
  }
})
options(figcap.prefix = "Figure", figcap.sep = ":", figcap.prefix.highlight = "**")

tabRef <- local({
  tag <- numeric()
  created <- logical()
  used <- logical()
  function(label, caption, prefix = options("tabcap.prefix"), 
           sep = options("tabcap.sep"), prefix.highlight = options("tabcap.prefix.highlight")) {
    i <- which(names(tag) == label)
    if (length(i) == 0) {
      i <- length(tag) + 1
      tag <<- c(tag, i)
      names(tag)[length(tag)] <<- label
      used <<- c(used, FALSE)
      names(used)[length(used)] <<- label
      created <<- c(created, FALSE)
      names(created)[length(created)] <<- label
    }
    if (!missing(caption)) {
      created[label] <<- TRUE
      paste0(prefix.highlight, prefix, " ", i, sep, prefix.highlight, 
             " ", caption)
    } else {
      used[label] <<- TRUE
      paste(prefix, tag[label])
    }
  }
})
options(tabcap.prefix = "Table", tabcap.sep = ":", tabcap.prefix.highlight = "**")

source("prepareSCRNAData.R")
source("prepareSCRNADataSet.R")
source("plotFunctions.R")
```

```{r data, include=FALSE, cache=TRUE}
sampleTable<-data.frame(Sample=c("S1", "S2", "S3"),
                        File=paste0("Z:/shengq1/20180214_scRNABatchQC/", c("S1.csv", "S2.csv", "S3.csv")))
  
sces<-prepareSCRNADataSet(sampleTable)
  
if(length(sces) < 9){
  scolors=brewer.pal(length(sces), "Set1")
}else{
  scolors=rainbow(length(sces))
}
names(scolors)<-names(sces)

```

## QC for single cell RNA-seq data

### 1. Technical features

```{r dis_total_counts,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("density_total_counts", "Distribution of total counts")}
print(plotDensity(sces,"total_counts", "Total counts", scolors))
```

```{r dis_total_feature,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("density_total_features", "Distribution of total features")}
print(plotDensity(sces,"total_features", "Total features", scolors))
```

```{r pct_counts_Mt,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("density_pct_counts_Mt", "Distribution of mitochondrial RNA")}
print(plotDensity(sces,"pct_counts_Mt", "PCT mitochondrial counts", scolors))
```

```{r gene_count_distribution,echo=FALSE,results='asis', warning=FALSE, fig.align="center", fig.cap=figRef("gene_count_distribution", "Features (x) and percentage of total counts (Y)")}
print(plotGeneCountDistribution(sces, scolors))
```

```{r AveCountVSNumberOfCells,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("AveCountVSNumberOfCells", "Average expression (x) and number of cells detect this gene (Y)")}
print(plotAveCountVSNumberOfCells(sces, scolors))
```

```{r plotVarianceTrend,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("VarianceTrend", "Mean-variance trend")}
print(plotVarianceTrend(sces, scolors))
```

```{r log10_total_counts,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("log10_total_counts", "log10(total counts)")}
print(plotMultiSamplesOneExplanatoryVariables(sces, scolors, "log10_total_counts", "log10(total counts)"))
```

```{r log10_total_features,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("log10_total_features", "log10(total features)")}
print(plotMultiSamplesOneExplanatoryVariables(sces, scolors, "log10_total_features", "log10(total features)"))
```

```{r log10_total_counts_Mt,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("log10_total_counts_Mt", "log10(total MT counts)")}
print(plotMultiSamplesOneExplanatoryVariables(sces, scolors, "log10_total_counts_Mt", "log10(total mitochondrial counts)"))
```

### 2. Sample expression similarity 

### 3. Bilogical similarity

```{r biologicalSimilarityHighVariableGenes,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("biologicalSimilarityHighVariableGenes", "Biological variation of most varied genes"), fig.height=10}
  plotBiologicalSimilarity(sces, objectName="hvg", filterName="FDR", valueName="bio")
```

```{r keggHighVariableGenes,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("biologicalSimilarityHighVariableGenes", "Biological variation of most varied genes"), fig.height=10}

wg<-WebGestaltR(enrichMethod="ORA",organism="mmusculus",
            enrichDatabase="pathway_KEGG",interestGene=hvgenes,
            interestGeneType="genesymbol",referenceSet="genome",
            is.output=FALSE)


```{r biologicalSimilarityPC1Genes,echo=FALSE,results='asis', fig.align="center", fig.cap=figRef("biologicalSimilarityPC1Genes", "Biological variation of PC1 genes"), fig.height=10}
pc1genes<-plotBiologicalSimilarity(sces, objectName="pc1genes", filterName="adj.P.Val", valueName="logFC")
wg<-WebGestaltR(enrichMethod="ORA",organism="mmusculus",
            enrichDatabase="pathway_KEGG",interestGene=pc1genes,
            interestGeneType="genesymbol",referenceSet="genome",
            is.output=FALSE)
```

